tests/summ_ai_api/summ_ai_test.py

```diff
@@ -222,3 +232,142 @@ async def test_summ_ai_error_handling(
+
+
+def test_check_rate_limit_exceeded(settings, caplog):
+    """
+    Test for check_rate_limit_exceedet method. Tests if return is True when response-status = 429, otherwise False
+
+    :param settings: The Django settings
+    :type settings: :fixture:`settings`
+
+    :param caplog: Fixture for asserting log messages in tests (see :fixture:`pytest:caplog`)
+    :type caplog: ~_pytest.logging.LogCaptureFixture
+    """
+
+    assert (
+        SummAiApiClient(MockedRequest(), "form_dummy").check_rate_limit_exceeded(200)
+        is False
+    ), "if response-status != 429, check-rate-limit should not return True"
+    assert (
+        SummAiApiClient(MockedRequest(), "form_dummy").check_rate_limit_exceeded(429)
+        is True
+    ), "if response-status == 429, check-rate-limit should return True"
+    for record in caplog.records:
+        assert (
+            "SUMM.AI translation of is waiting for "
+            + str(settings.SUMM_AI_WAIT_FOR_NEXT_REQUEST_AFTER_LIMIT_EXCEEDED)
+            + "s"
+            in record.message
+        ), "after check_rate_limit_exceeded(429) logging entry should appear"
+    assert (
+        SummAiApiClient(MockedRequest(), "form_dummy").check_rate_limit_exceeded(500)
+        is False
+    ), "if response-status != 429, check-rate-limit should not return True"
+
+
+# pylint: disable=too-few-public-methods
+class MockedRequest:
+    """
+    Helper class mocking request, used for creating a SummAiApiclient instance. Region-property is needed therefore.
+
+    """
+
+    def __init__(self):
+        self.region = 1
+        self.data = {"dummy": "dummy request"}
+
+
+def test_patient_task_queue_normal_deque():
+    """
+    Test for PatientTaskQueue Class. Tests that a last-in/first out (lifo) deque-queue is build from a task-list
+    Checks that next() removes one element each from the queue
+    """
+
+    tasks = [1, 2, 3]
+    task_generator = PatientTaskQueue(tasks, 1)
+
+    # we check that list comes reverse (lifo-queue)
+    for i in range(len(task_generator)):
+        assert next(task_generator) == tasks[len(tasks) - 1 - i]
+
+    assert len(task_generator) == 0
+    # we check that next() did remove elements from the queue
+    with pytest.raises(StopIteration):
+        next(task_generator)
+
+
+async def test_patient_task_queue_hit_rate_limit():
+    """
+    Test for PatientTaskQueue Class. It tests that when the maximum number of parallel requests
+    to the summ_ai_api has been exceeded (rate limit exceeded), a defined time is waited until the next task object is returned via next().
+    """
+
+    tasks = [1, 2, 3]
+    # for the test, waiting-time is reduced to 0.5s, default is 60s defined in settings.SUMM_AI_WAIT_FOR_NEXT_REQUEST_AFTER_LIMIT_EXCEEDED
+    task_generator = PatientTaskQueue(tasks, 0.5)
+
+    # Rate limit hit
+    task = next(task_generator)
+    assert task in tasks
+    # we call out a reach the rate limit
+    task_generator.hit_rate_limit(task)
```

Comment:
We intended to write tests for (nearly) all abstraction levels, both high level, overarching ones, similar to those that currently make up the bulk of all tests, as well low level ones, testing only single components, ensuring the origin of a breaking change can be quickly located.

Reply:
Hi @Timo, 
Thanks for the very good review, the helpful hints and especially your close look at the logic! I have never met a developer who did this so well.
The git issues were almost certainly me screwing up (again). I'm continuing to make a genuine effort and ask for your utmost indulgence - I'm quite annoyed with myself.
As far as the testing discussion is concerned, I would like to take up the cudgels for extending unit tests and also testing encapsulated, very simple methods whenever we touch code. The functional tests (e.g. API tests) have cost us an incredible amount of time to fix. When functional tests fail, it is almost impossible to know where exactly the problem is without knowing the full context. As projects become more complex and the number of developers involved increases, this becomes a bigger and bigger problem. I have made the painful experience that a larger test coverage (only) of functional tests created far more problems than it solved. 
Should we test (in Python, after all, only defined by usage) private methods? Actually not useful. However, where we can reduce complexity by outsourcing logic to our own methods, we should do so anyway - my opinion. The extra effort will benefit us in two ways: we will be able to test much more granularly and see more and more where a problem occurs, we will reduce the complexity of methods and generate even more readable code (production as well as test code). Where new code is written test driven, it will no longer be necessary/meaningful to test private methods.
Furthermore, our attention should be focused on reducing test execution times. The current waiting times cost us a lot of productivity and will slow down the acceptance of TDD. In my view, this includes keeping tests with database access to a minimum and using a memory DB as the test database. 
As @Peter noted, this could become a discussion :-).